### PR TITLE
Cleanup markers on close & consider marker reason

### PR DESCRIPTION
**Pull request by planger**
With this change, we clean-up all markers on close of a diagram widget but only keep by default those markers that have been added by a live validation.

https://github.com/eclipse-glsp/glsp/issues/980

For review, please see [the comment on which changes belong together](https://github.com/eclipse-glsp/glsp/issues/980#issuecomment-1516867263).

Change-Id: I62e515505a4a4afeed1d187393a471073d22a089

### CODE REVIEW

**Comment by planger** [comment]
Build failure is expected as it depends on the protocol change:
```
@eclipse-glsp/theia-integration: src/browser/diagram/theia-marker-manager.ts(16,72): error TS2305: Module '"@eclipse-glsp/client/lib"' has no exported member 'MarkersReason'.
```

File: packages/theia-integration/src/browser/diagram/theia-marker-manager.ts
Comment (tortmayr) [inline packages/theia-integration/src/browser/diagram/theia-marker-manager.ts:45]: Is there a particular reason why we don add the `reason` property directly to the `Marker` interface in the GLSP client packages?

File: packages/theia-integration/src/browser/diagram/theia-marker-manager.ts
Comment (tortmayr) [inline packages/theia-integration/src/browser/diagram/theia-marker-manager.ts:72]: Nitpick: Could also be done as a oneline with the filter API
```ts
 getMarkerByOrigin(origin?: string): ValidationMarker[] {
       return Array.from(this.diagnotic2marker.values()).filter(marker => marker.reason === origin);
    }
```

File: packages/theia-integration/src/browser/diagram/theia-marker-manager.ts
Comment (tortmayr) [inline packages/theia-integration/src/browser/diagram/theia-marker-manager.ts:73]: Nitpick:
You could use `forEach` here instead of iterating over the keys. This way you don't have to retrieve the value from the map and can avoid the `undefined` check:

```ts
    this.diagnotic2marker.forEach((marker,diagnostic)=>{
            if (marker.reason === origin){
                diagnostics.push(diagnostic);
            }
        });
```

File: packages/theia-integration/src/browser/diagram/theia-marker-manager.ts
Comment (tortmayr) [inline packages/theia-integration/src/browser/diagram/theia-marker-manager.ts:83]: Nitpick: Same as above,`forEach` could be used here.

File: packages/theia-integration/src/browser/diagram/theia-marker-manager.ts
Comment (tortmayr) [inline packages/theia-integration/src/browser/diagram/theia-marker-manager.ts:186]: I suggest the following to get rid of the ! operator:
```ts
  const resourceUri= getDiagramWidget(widget)?.getResourceUri();
  if (resourceUri) {
      this.clearMarkers(resourceUri, this.markerReasonsToKeep);
  }
```

**Review by tortmayr — CHANGES_REQUESTED** [review]
Thanks Philip, support for live validation is definitely a nice feature addition.
I did not have time to test the changes yet and only did a preliminary code review for now.

File: packages/theia-integration/src/browser/diagram/theia-marker-manager.ts
Comment (planger) [reply packages/theia-integration/src/browser/diagram/theia-marker-manager.ts:45]: Yes, sort of... in my first version I even had it this way. But I opted to only mention the `reason` in the `SetMarkersAction` to make it more clear and easier to handle that each `SetMarkersAction` is meant to replace all markers that of a previous `SetMarkersAction` with the same `reason`.
If we'd put the `reason` in the `Marker` we may receive a set of markers with mixed reasons, which would complicate the handling on the client to remove the correct markers from previous `SetMarkersAction`.
But I agree, it is a bit cumbersome here that we need to remember it by extending the interface.

**Comment by planger** [comment]
Thank you for the swift and excellent review! I've applied your suggestions in 3836bb7.

**Review by tortmayr — APPROVED** [review]
LGTM! Once the client change is merged we can update the locked version and merge this aswell.

**Comment by tortmayr** [comment]
Idea for a follow-up issue/feature:
I think it would be nice if we also would provide the option to enable/disable live validation via the Theia settings.

**Comment by planger** [comment]
> I think it would be nice if we also would provide the option to enable/disable live validation via the Theia settings.

Good idea, I've tracked it here for now: https://github.com/eclipse-glsp/glsp/issues/980#issuecomment-1517861283

@tortmayr Also, can you please re-check f4eab49 in which I've upgraded the `yarn.lock`? I ran `yarn upgrade:next` in the theia integration repo, but this resulted in quite a lot of updates, which is why I think it makes sense if you double-check.
If this is ok, feel free to directly squash and merge. Thank you!

**Pull request merged**
